src/meta/src/stream/fragmenter.rs

```diff
@@ -141,9 +147,64 @@ where
             .collect::<Result<BTreeMap<_, _>>>()
     }
 
+    /// Do some dirty rewrites on meta. Currently, it will split stateful operators into two
+    /// fragments.
+    fn rewrite_stream_node(&self, stream_node: StreamNode) -> Result<StreamNode> {
+        self.rewrite_stream_node_inner(stream_node, false)
+    }
+
+    fn rewrite_stream_node_inner(
+        &self,
+        stream_node: StreamNode,
+        insert_exchange_flag: bool,
+    ) -> Result<StreamNode> {
+        let mut inputs = vec![];
+
+        for child_node in stream_node.input {
+            let input = match child_node.get_node()? {
+                // For stateful operators, set `exchange_flag = true`. If it's already true, force
+                // add an exchange.
+                Node::HashAggNode(_) | Node::HashJoinNode(_) => {
+                    // We didn't make `fields` available on Java frontend yet, so we check if schema
+                    // is available (by `child_node.fields.is_empty()`) before deciding to do the
+                    // rewrite.
+                    if insert_exchange_flag && !child_node.fields.is_empty() {
+                        let child_node = self.rewrite_stream_node_inner(child_node, false)?;
+
+                        let strategy = DispatchStrategy {
+                            r#type: DispatcherType::NoShuffle.into(),
+                            column_indices: vec![],
+                        };
+
+                        StreamNode {
+                            pk_indices: child_node.pk_indices.clone(),
+                            fields: child_node.fields.clone(),
+                            node: Some(Node::ExchangeNode(ExchangeNode {
+                                strategy: Some(strategy.clone()),
+                            })),
+                            operator_id: 10000000 + child_node.operator_id,
```

Comment:
😂

Reply:
This id is not important -- it won't be stored in meta or used by something else, and it's not even globally unique. So it's okay to assign such value here.